CONTRIBUTING.md

```diff
@@ -0,0 +1,179 @@
+```
+
+To run the unit tests in isolation:
+```
+npm run unit
+```
+
+When done you want to test a new branch or internal test interview files have been updated, clean up before setting up again:
+```
+npm run takedown
+```
+
+# Very general architecture of files and folders
+
+An honest look at our current project architecture - some of our files and folders, what they do, and how they interact.
+
+## Logic architecture
+
+### .feature files
+
+The `.feature` files are written in Gherkin, a syntax cucumber uses. The "code" in there relies on the functions set up in `./lib/steps.js`.
+
+```js
+// The test_something.feature file step
+Then I sign
+
+// relies on the code in steps.js
+Then('I sign', { timeout: -1 }, async () => {
+  return wrapPromiseWithTimeout(
+    scope.steps.sign( scope ),
+    scope.timeout
+  );
+});
+```
+
+### steps.js
+
+`./lib/steps.js` defines all the Gherkin sentences that our users (the developers) can write. See [cucumber documentation on defining steps](https://cucumber.io/docs/cucumber/step-definitions/?lang=javascript). We sometimes use "cucumber expressions" and, but often use regular expressions as we often want to be more flexible about the user's input.
+
+The file handles things like:
+* Creating state
+* Defining steps for the users/developers
+* Defining steps for internal testing, like testing for appropriate errors and error messages
+* Finishing a Scenario
+* Finishing all tests
+
+### scope.js
+
+`./lib/steps.js` makes heavy use of `./lib/scope.js`, where most of the action happens. We have decided to mostly keep it in one file to avoid bouncing back and forth between files.
+
+Many `./lib/steps.js` steps make use of the `scope.steps` object. It is there to avoid duplicating code between steps.
+
+The file hanldes things like:
+* Finding form fields on a page
+* Filling out form fields
+* Generating random data for form fields
+* Waiting for pages to load
+* Checking for server restarts
+* Generating test report content
+
+## Logging
+
+`./lib/utils/log.js` takes care of logging information of various kinds (normal, debug, error) to the console in ways we find useful. We should probably find a logging library instead, but this is what we have for now. An important feature is that it ensures the messages clearly belong to ALKiln and not to some other process.
+
+## session_vas.js
+
+The `./lib/utils/session_var.js` file keeps track of what you might otherwise think of as environment variables. We like to think of them as constants, but some of them do need go through functions, so for consistency we get them all through functions. Also because it was easier to test them when they are functions.
+
+## Setup and takedown logic
+
+All of the setup and takedown logic is in the `./lib/docassemble` directory. Not a great name. It is named after the docassemble API code that is in there. Also, setup and takedown interact very closely with docassemble and the docassemble server.
```

Comment:
```suggestion
All of the setup and takedown logic is in the `./lib/docassemble` directory. It is named after the docassemble API code that is in there. Also, setup and takedown interact very closely with docassemble and the docassemble server.
```

IMO it's a fine name, lol

Reply:
I'd like to be clear about what we see as current down-sides to the codebase. I'd rather discuss the flaws. What are the cons you see to that?

Reply:
My main thought here was that you say it's not a great name, and then describe why it's not that bad of a name. It might not be perfect, but no name is going to match everyone's mental model; that's part of learning a new code base, learning what existing terms are for things.

Speaking to flaws is fine, but from a teaching perspective (which this document is trying to teach folks about the code base), pointing out bad names isn't really helpful. I think the next sentence about this folder being a bit messy is good (it gives folks expectations about what they can expect if they read it), and the footnote about the "misuse of cucumber" is worded a bit more negatively than I would have personally put it, but is a really useful point to make, that we aren't trying to do BDD, but are more using it for the "English as code" features.